3p/ampcontext.js

```
@@ -0,0 +1,203 @@
+/**
+ * Copyright 2015 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import './polyfills';
+import {
+  dev,
+  user,
+} from '../src/log';
+import {IframeMessagingClient} from './iframe-messaging-client';
+
+/**
+   Enum for the different postmessage types for the window.context
+   postmess api.
+*/
+export const MessageType_ = {
+  SEND_EMBED_STATE: 'send-embed-state',
+  EMBED_STATE: 'embed-state',
+  SEND_EMBED_CONTEXT: 'send-embed-context',
+  EMBED_CONTEXT: 'embed-context',
+  SEND_INTERSECTIONS: 'send-intersections',
+  INTERSECTION: 'intersection',
+  EMBED_SIZE: 'embed-size',
+  EMBED_SIZE_CHANGED: 'embed-size-changed',
+  EMBED_SIZE_DENIED: 'embed-size-denied',
+};
+
+export class AmpContext extends IframeMessagingClient {
+
+  /**
+   *  @param {Window} win The window that the instance is built inside.
+   */
+  constructor(win) {
+    super(win);
+    this.setupMetadata_();
+    this.ampWindow = this.getHostWindow();
+  }
+
+  /** @override */
+  getHostWindow() {
+    const sentinelMatch = this.sentinel.match(/((\d+)-\d+)/);
+    if (sentinelMatch) {
+      this.depth = Number(sentinelMatch[2]);
+      this.ancestors = [];
+      for (let win = this.win_; win && win != win.parent; win = win.parent) {
+        // Add window keeping the top-most one at the front.
+        this.ancestors.unshift(win.parent);
+      }
+      return this.ancestors[this.depth];
+    } else {
```

nitpick: can we reverse the if-else? i.e., fail early in the code and save some indentation.